[Refactoring 1] Introducing abstract liquidity object

fleupold opened the pull request:
Part of #194

*This is a proposal for how we can separate liquidity sources and solver handling logic to allow the two to grow more independently. Read in the context of the three PRs in this stack*

This PR defines an abstract Liquidity type, which can be interpreted by the solver(s). For now this datatype is shared between the two solvers (naive & http). If we see more solvers evolve in the future (with fundamentally different order types) it might make sense to have different liquidity types, one for each solver. This way, we could e.g. still model a Curve pool as limit orders for more naive solvers, but e.g. provide native support in the http_solver.

On the other hand, we have many liquidity sources (e.g. Offchain Orderbook, SCP RFQ system, Uniswap, SushiSwap. Balancer, Curve, Compound cToken converter, etc...). Each liquidity source yields a lot of abstract liquidity and provides a callback handler into how a specific "match" of this liquidity translates into trades and interactions.

As an example, if we have a user selling GNO for ETH, our source would provide a LimitOrder object selling GNO for WETH. In the settlement handler we would return a GNO/WETH trade **and** an interaction to unwrap the amount of WETH bought into ETH (we might build additional logic towards how to order/merge interactions in the end to avoid e.g. doing multiple unwrap calls).

Similarly an AMM liquidity source (e.g. Uniswap or Sushiswap) would yield an AMM liquidity object. In the settlement handler it could perform things such as approving the settlement contract if necessary.

An external market maker's RFQ system can be implemented as a liquidity source with a list of partially matchable limit orders that get matched via a 0x bridge (opposed to native GPv2 EOA orders).

PredictionMarket outcome tokens or other "split and mergeable" tokens (e.g. LP tokens) would become a new liquidity subtype (once the solvers support it) with potentially many sources.

### Test Plan
Unit tests at the end of the PR stack

fleupold committed:
[Refactoring 1] Introducing abstract liquidity object

vkgnosis commented on solver/src/liquidity.rs:
```diff
@@ -0,0 +1,75 @@
+    ///
+    async fn get_liquidity(
+        &self,
+        liquidity_so_far: impl Iterator<Item = &Liquidity> + Send + Sync + 'async_trait,
```
Shouldn't need Sync

vkgnosis commented on solver/src/liquidity.rs:
```diff
@@ -0,0 +1,75 @@
+
+use crate::settlement;
+
+///
+/// Defines the different types of liquidity our solvers support
+///
```
Why surround doc comment with empty lines? Not important but I prefer just having the comment.

vkgnosis commented on solver/src/liquidity.rs:
```diff
@@ -0,0 +1,75 @@
+///
+#[derive(Clone)]
+pub struct AmmOrder {
+    pub tokens: (H160, H160),
```
Should be TokenPair

josojo commented:
Cool. Yes, your suggestion makes a ton of sense to me. Like it.

Nit:
>If we see more solvers evolve in the future (with fundamentally different order types) it might make sense to have different liquidity types, one for each solver.

Since we don't wanna pull and store the information of one liquidity source several times for each solver, I would introduce another trait: `representLiquidityForSolver`, that would transform one Liquidity object of the Liquidity-Enum and destruct it into a good representation for the solver. For each liquidity, we can have several `representLiquidityForSolver` traits implemented.

E.g:
```
Liquidity: Curve - AMM
```
```
Implementation 1: representLiquidityForSolver -> destructs the AMM into linear orders
```
```
Implementation 1: representLiquidityForSolver -> just takes the reserve values and sends it to the solver.
```

But I think the introduction of this trait can also happen later.

fleupold committed:
Removing LiquiditySource trait

fleupold committed:
comments

fleupold commented:
I addressed the comments and following up on the discussion in #194 I removed the LiquiditySource trait but am leaving the liquidity type as an enum for now.

fleupold committed:
Merge branch 'main' into liquidity_refactoring_1
